### PR TITLE
feat: use urban typology for activity chain matching

### PR DESCRIPTION
**Pull request by aburianne**
ENTD (national mobility survey) can be used to generate synthetic population (with Synpop) over any metropolitan french territory. It is mostly used whether there are no available EMD or EDGT (local mobility survey with higer sampling). 

ENTD is (main method) filtered by departments (i.e.: Gironde-33, Paris-75) composing the survey area, leaving us with critical low sampling issues. Therefore we aim to change approach on sampling and parameters inputs for agents mobility behaviors.

Our proposal is to keep all of metropolitan France (NO department filter) and to apply a typology based on 4 classes :

- central city 
- suburbs 
- isolated city
- rural / countryside

Doing so, we assume same mobility behaviors patterns within classes, i.e. : Lyon, Bordeaux, Marseille... and significant deviation between classes. 

This 4 classes typology is computed and __[open source distributed by INSEE](https://www.insee.fr/fr/information/4802589)__. It is refered to as "Unités Urbaines" (UU). INSEE internal main use of this dataset is to adapt statistics methods to context. The data comes as Microsoft Excel (.xlsx) worksheet with a 2010 and a 2020 version with official geographical code milleisim options.

The UU typology being availble both in ENTD and official geographical codes for all of metropolitan France it is suitable for the process. 

- [x] review code
- [x] set option name and mecanism ( urban_units or any other relevant name)
- [x] write and validate test (new fake data set of UU and hascodes)
- [x] changes for other territories? (Corisca, Toulon, Lyon, Nantes...)
- [x] write documentation
- [x] do we share the jupyter notebook with ENTD filtering methods comparison for reproductible science, so people can evaluate on any french territory?

### CODE REVIEW

File: matsim/runtime/eqasim.py
Comment (aburianne) [inline matsim/runtime/eqasim.py:1]: Changed for use with Tarek developments on mode choice (but not shure of what version to keep)

File: data/spatial/uu.py
Comment (aburianne) [inline data/spatial/uu.py:1]: had issue with excel file when reading (fixes by manually open / save / close in excel). Maybe a different library or file fix?

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,4 @@ data/lyon_2015
 .vscode
 
 config_local_*.yml
+config_bordeaux*.yml
```
Comment (sebhoerl) [inline .gitignore:11]: Remove

File: config.yml [hunk 2]
```diff
@@ -7,7 +7,7 @@ working_directory: cache
 # This section defines which parts of the pipeline should be run
 run:
   - synthesis.output # To create the output population in the output_path (see below)
-  #- matsim.output # Uncomment, if you want to run the full simulation (you'll need Java for that)
+  - matsim.output # Uncomment, if you want to run the full simulation (you'll need Java for that)
```
Comment (sebhoerl) [inline config.yml:10]: Reset

File: config_bordeaux_0.1.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+## Synthetic population pipeline for Île-de-France
```
Comment (sebhoerl) [inline config_bordeaux_0.1.yml:1]: Remove file

File: data/census/cleaned.py [hunk 4]
```diff
@@ -12,9 +12,11 @@
 def configure(context):
     context.stage("data.census.raw")
     context.stage("data.spatial.codes")
+    context.stage("data.spatial.uu")
 
 def execute(context):
     df = context.stage("data.census.raw")
+    uu = context.stage("data.spatial.uu")
```
Comment (sebhoerl) [inline data/census/cleaned.py:19]: change to df_urban_type

File: data/census/cleaned.py [hunk 5]
```diff
@@ -111,12 +113,18 @@ def execute(context):
     # Consumption units
     df = pd.merge(df, hts.calculate_consumption_units(df), on = "household_id")
 
+    # UU match
```
Comment (sebhoerl) [inline data/census/cleaned.py:116]: "Impute urban type"

File: data/census/cleaned.py [hunk 6]
```diff
@@ -111,12 +113,18 @@ def execute(context):
...
+    df = pd.merge(df,uu,on="commune_id",how="left")
+    df.loc[df["commune_id"] == "undefined","type_uu"] = "rural"
```
Comment (sebhoerl) [inline data/census/cleaned.py:118]: I would call "rural" as "none", because it does not have an "urban type"

File: data/census/cleaned.py [hunk 7]
```diff
@@ -111,12 +113,18 @@ def execute(context):
...
+
     return df[[
         "person_id", "household_id", "weight",
         "iris_id", "commune_id", "departement_id",
         "age", "sex", "couple",
         "commute_mode", "employed",
         "studies", "number_of_vehicles", "household_size",
         "work_outside_region", "education_outside_region",
-        "consumption_units", "socioprofessional_class"
+        "consumption_units", "socioprofessional_class","type_uu"
```
Comment (sebhoerl) [inline data/census/cleaned.py:129]: Rename column to urban_type

File: data/hts/entd/cleaned.py [hunk 8]
```diff
@@ -115,6 +115,11 @@ def execute(context):
     df_trips["origin_departement_id"] = df_trips["V2_MORIDEP"].fillna("undefined").astype("category")
     df_trips["destination_departement_id"] = df_trips["V2_MDESDEP"].fillna("undefined").astype("category")
 
+    # Clean unités urbaines    
+    df_households["type_uu"] = df_households["numcom_UU2010"].replace({"B":"suburb","C":"central_city","I":"isolated_city","R":"rural"})
+    assert np.all(~df_households["type_uu"].isna())
```
Comment (sebhoerl) [inline data/hts/entd/cleaned.py:120]: column urban_type

File: data/hts/hts.py [hunk 9]
```diff
@@ -181,7 +181,7 @@ def check_activity_types(df_trips):
     error_count = np.count_nonzero(f)
     print("Trips with inconsistent activity types: %d" % error_count)
 
-    return error_count == 0
+    return error_count == 1
```
Comment (sebhoerl) [inline data/hts/hts.py:184]: Set back to zero, fix if it gives an error

File: data/hts/hts.py [hunk 10]
```diff
@@ -222,7 +222,7 @@ def calculate_consumption_units(df_persons):
 HOUSEHOLD_COLUMNS = [
     "household_id", "household_weight", "household_size",
     "number_of_vehicles", "number_of_bikes", "departement_id",
-    "consumption_units", # "income_class"
+    "consumption_units", "type_uu" # "income_class"
```
Comment (sebhoerl) [inline data/hts/hts.py:225]: column name

File: matsim/runtime/eqasim.py
Comment (sebhoerl) [reply matsim/runtime/eqasim.py:1]: let's update to latest develop

**Review by sebhoerl — CHANGES_REQUESTED** [review]

**Comment by sebhoerl** [comment]
- Cleaned up the code
- I propose to name the attributes `urban_type` and then have either `suburb`, `center_city`, `isolated_city`, but all the remaining zones are `none`

**Review by sebhoerl — APPROVED** [review]

**Comment by sebhoerl** [comment]
To use the urban type one needs now to
- set `use_urban_type: true` in the configuration (this imputes the information)
- set `matching_attributes: ["urban_type", "*default*"]` this appends the `urban_type` to the default matching attributes

**Pull request merged**